### ISSUE TITLE
Migrating this project to Gitlab.com

### ISSUE THREAD

**Issue by nicholasklick** [gitlabhq/terraform-provider-gitlab#445]
There is interest from the Gitlab org to migrate this project to our .COM site.
Now that we have shipped 3.0.0 it seems like a relevant time to migrate.

We need to investigate:

- [ ] Interest or concerns the current community may have around this move
- [ ] Ability to mirror in such a way that continues to work with the Hashicorp registry 
- [ ] Migrating current issues
- [ ] Migrate current PRs
- [.] Migrate actions => CI jobs

**Comment by nicholasklick** [comment]
@roidelapluie @armsnyder any opinions re this?

**Comment by roidelapluie** [comment]
I am quite against this if we need a mirror on GitHub.

Why? GitHub does not allow you to turn off pull requests. We would still get new pull requests. It would be unfriendly to each time ask people to go to gitlab.com to make their mr. A bot does not seem better.

As long as hashicorp only supports GitHub, I'd prefer to stay here.

**Comment by nicholasklick** [comment]
@roidelapluie I definitely value your opinion here.

Though it seems like I can halt new PRs, issues and comments from anyone not a maintainer here: https://github.com/gitlabhq/terraform-provider-gitlab/settings/interaction_limits.

We would communicate clearly that the repo was a mirror similar to: https://github.com/gitlabhq/omnibus-gitlab

Do you have other concerns besides locking PRs and comments etc?

**Comment by roidelapluie** [comment]
mmh  that sounds like an abuse of moderation but yes that would then work.

**Comment by nicholasklick** [comment]
@roidelapluie thanks for your thoughts here. I am not sure it is really an abuse of moderation given we are trying redirect conversation to a specific venue: gitlab.com. All would be welcome to contribute on the gitlab.com platform. I don't want to restrict transparency nor collaboration rather ensure there is a single source of truth for everyone to contribute to.

**Comment by roidelapluie** [comment]
:+1:

**Comment by armsnyder** [comment]
In my opinion, the project should live wherever the larger community of interested and potential contributors is. I'm unfamiliar with the gitlab.com community, but I would say that if you think there will be more active community engagement there, then we should go for it. Otherwise, the complexity of mirroring may not be worth it. Wouldn't releases have to be done in github still, for Terraform Registry to detect them?

**Comment by roidelapluie** [comment]
Note: We also need to change the go mod path

**Comment by CraigGardener** [comment]
Regardless of whether the source of truth is github.com or gitlab.com, it would be helpful to have a copy (and clear guidance/docs) on both.

Whether the gitlab repo is mirrored to the github repo or vice versa, the documentation and project descriptions should ideally include a link to at least the primary repo in order to provide clarification for potential contributors and those interested in the development.


Personally, I'd expect to find a project maintained by GitLab on gitlab.com

**Comment by nicholasklick** [comment]
@armsnyder @roidelapluie This issue has come up again.

We are getting many customers and internal users who are confused and want to create GitLab issues for the repository.

We are debating what to do and would welcome any additional input you have.

**Comment by armsnyder** [comment]
@nicholasklick How does publishing to the registry work today? Is it triggered automatically when you create a release in GitHub? It seems like we can't get out of making releases from a GitHub project.

Is there a way to mirror issues so that folks can create issues in gitlab.com and they transfer here?

**Comment by CraigGardener** [comment]
Potential solution… locked GitHub repo (with DECENT DOCUMENTATION about
where to find the source) that a GItLab CI job pushes to in order to create
a release (to keep terraform happy). All dev/issues takes place in GitLab.

It seems bizarre for a competitor of GitHub to primarily use GitHub for
anything. Should it not be a last resort if GitLab claims feature-parity?

My two cents:
- Shift the PRIMARY repo to GitLab.
- Use GitHub only where necessary (to create releases for TF as it would
appear) - if that’s a separate “release-ready” repo that’s only pushed to
by a CI job, what’s the problem?
- work with Terraform to make Gitlab a first class citizen within the TF
ecosystem in the same way that GitHub is

**Comment by armsnyder** [comment]
I think the main limitation is in the Terraform registry, which can only pull release artifacts from GitHub. (@nicholasklick can confirm since he does the releases.)

**Comment by armsnyder** [comment]
The "Report an issue" button in the registry takes you to the GitHub project. I think the link would be broken if we moved the project and disabled issues in the GitHub mirror. https://registry.terraform.io/providers/gitlabhq/gitlab/latest/docs

**Comment by nagyv** [comment]
Moving the repo to GitLab would be of great help to the core GitLab team and many of the non-contributing users. With the team, we try to plan every milestone the issues we want to deliver and the community PRs we want to support to get merged. Today, this requires us linking issues and we can't even track progress together with the rest of our milestone work.

----

As we are in contact with Hashicorp, we learned their plans around potentially supporting GitLab as a source repo for the public registry, and it's pretty low in their priority list (it might happen at in the next financial year).

We can build out many automations. Based on the discussion until now, I see the following:

- mirror the code repo from GitLab to GitHub
- automate releases on GitHub, triggered by GitLab
- automatically close any issues opened on GitHub with a note, and perhaps open an issue in GitLab using a bot account

Do you see any other areas where moving to GitLab might be an issue?

**Comment by armsnyder** [comment]
* Migrate the existing GitHub pull requests and issues to the new GitLab source (hopefully including labels)
* Find a way of making it clear that the GitHub project is a read-only mirror. (I don't think we could modify the README if it's a mirror, right?)
* Before officially cutting over to using GitLab as the source, do a full run-through of mirroring the GitLab project to a mockup repository in GitHub and test the release automation, in order to minimize disruption.

**Comment by roidelapluie** [comment]
I propose that we DELETE the code in this repository and disable issues.


This will make sure that NO PULL REQUEST come on github and we can focus on gitlab.com


We would just need to push the TAGS to github.com. We do not need to push the branches to github.com. The TAGS will contain the full code but not be related to branches.


Therefore, the MAIN branch on github.com can contain a single README file that explains that the code is on github.

**Comment by kgoossens** [comment]
> I don't think we could modify the README if it's a mirror, right?

Should be possible to add it to the README in both
e.g. stating that the repo on gitlab.com is the 'primary' repository and github.com contains a 'secondary' readonly version. This kind of statement is valid to exist in both registries? 

/my2cents

**Comment by roidelapluie** [comment]
I am against having a mirror because we can not disable pull requests and it will be annoying. I think we should go out of github, put the branches on gitlab and only push tags to github.

**Comment by nicholasklick** [comment]
> automatically close any issues opened on GitHub with a note

@nagyv here is an example of a GitHub action that can accomplish this: https://github.com/marketplace/actions/repo-lockdown

**Comment by nicholasklick** [comment]
@armsnyder 

> How does publishing to the registry work today?

Currently it is a fairly manual process, but needs to be automated.

> It seems like we can't get out of making releases from a GitHub project and would still need GitHub workflows to create release artifacts.

This is true, we have asked Hashi about this and they plan to only support GitHub for release generation for the foreseeable future. 

> Is there a way to mirror issues so that folks can create issues in gitlab.com and they transfer here?

I think it would be best to block GitHub issues and point people to create GitLab issues instead to have a single source of truth.

**Comment by nicholasklick** [comment]
DRAFT migration plan: https://gitlab.com/gitlab-org/configure/general/-/issues/150

**Comment by nicholasklick** [comment]
@roidelapluie 

> I am against having a mirror because we can not disable pull requests and it will be annoying. I think we should go out of github, put the branches on gitlab and only push tags to github.

I tend to agree with this to not duplicate PRs and issues across platforms. GH issues and PRs should either be locked or be auto-closed via a bot.

**Comment by roidelapluie** [comment]
issues can be disabled, and we do not have to push code here. We can push a single readme and push the tags, making it very unlikely to receive pull requests in the first place.

**Comment by nagyv** [comment]
We plan to focus on the migration after the GitLab 15.0 release.

**Comment by timofurrer** [comment]
FYI: we are actively working on planning and probably soon-ish execute the migration of this repository over to GitLab. You can follow the process here: https://gitlab.com/groups/gitlab-org/configure/-/epics/10

I'll be posting information here to keep you all up to date. Please, as always, don't hesitate to reach out if you have any doubts, comments or anything else 🤝

**Comment by timofurrer** [comment]
**(maybe) Action Required**: In order for your user names to be matched correctly the public facing email addresses on GitHub and GitLab must match, as described [here](https://docs.gitlab.com/ee/user/project/import/github.html#prerequisites). If you want your user to correctly be referenced on the new GitLab repository, please make sure this works. You can also verify this by importing one of your projects from GitHub to GitLab and check if your user is assigned properly.

**Comment by PatrickRice-KSC** [comment]
@timofurrer  - Anything we can help with on the gitlab transition side? I was going to pick up some open issues for work, but figured it may be better to wait on  that until the transition is done.

**Comment by timofurrer** [comment]
@PatrickRice-KSC no - we are awaiting a few things internally. Hopefully we can start the migration early this week.

**Comment by PatrickRice-KSC** [comment]
Sounds good, just didn't want to leave all the work to you if there was something we could help with!

**Comment by timofurrer** [comment]
FYI: today early afternoon UTC we'll start migrating this repository over to GitLab. I'll keep you posted.

**Comment by timofurrer** [comment]
The repository has been migrated to https://gitlab.com/gitlab-org/terraform-provider-gitlab and we'll proceed with the migration outlined here: https://gitlab.com/gitlab-org/configure/general/-/issues/241#note_1152660972 🎉
